### PR TITLE
WebUSB

### PR DESCRIPTION
**Pull request by marcos-diaz**
Highlights of the PR:
- WebUSB USB descriptors.
- USB is initialized first, then the rest of the initializations, so everything can be logged through WebUSB if connected.
- Logging functions such as `info()`, `warning()` and `debug()` are a wrapper for `printf()`/UART and the WebUSB communication.
- Initial implementation of the Ctrl protocol for the Ctrl app. See `/docs/ctrl_protocol.md`.
- Tune options are now generalized as config changes that can be triggered both from the tune shorcuts and the Ctrl app.

### CODE REVIEW

File: docs/ctrl_protocol.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
+# Ctrl protocol
+
+## Introduction
+
+The Ctrl protocol is a minimalistic bidirectional communication protocol between the controller and the desktop app, using USB (WebUSB). For the purpose of exchanging data such as logging, configuration settings and profile mappings.
```
Comment (przymi) [inline docs/ctrl_protocol.md]: progressive / web app

File: docs/ctrl_protocol.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+The exchange is performed with discrete packages of at most 64 bytes.
+
+Both sides of the communication (USB host and USB device) are equally-levelled actors from the point of view of the protocol, both sides can initiate the communication (send a message) at any time.
+
+There is no acknowledge when receiving messages at protocol level, though some messages could trigger a response message from the other side. In such cases the response message is asynchronous, and can happen at any time later or not happen at all. Is up to the "receiver" to determine adequate timeout values.
+
+Padding zeros to complete the 64 bytes may be used. Therefore the protocol avoids using zero as indexes for any of its enums.
+
+## Package general structure
```
Comment (przymi) [inline docs/ctrl_protocol.md:17]: Redundancy / checksumming provided by the usb protocol

File: docs/ctrl_protocol.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+### Message type
+
+| Key | Index |
+| - | - |
+LOG | 1
+PROC | 2
+CONFIG_GET | 3
+CONFIG_SET | 4
+CONFIG_GIVE | 5
```
Comment (przymi) [inline docs/ctrl_protocol.md]: "give" is confusing. It's a return or notify

File: docs/ctrl_protocol.md [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+
+## Log message
+Direction: `Controller` -> `App`
+
+| Byte 0 | 1 | 2 | 3 | 4-63 |
+| - | - | - | - | - |
+| Protocol version | Device Id | Message type | Payload size | Payload
+|                  |           | LOG          | 1-60         | Log message
+
+## Proc message (procedure)
```
Comment (przymi) [inline docs/ctrl_protocol.md]: What is this used for? It appears to initiate different settings / calibration procedures. Does it then duplicate (parts of) the CONFIG or PROFILE _SET?

File: docs/ctrl_protocol.md
Comment (przymi) [inline docs/ctrl_protocol.md:1]: Some example / sequence chart would be nice.

File: src/bus.c [hunk 5]
```diff
@@ -95,14 +96,14 @@ uint8_t bus_spi_read_one(uint8_t cs, uint8_t reg) {
...
-    printf("INIT: I2C bus\n");
+    info("INIT: I2C bus\n");
     i2c_init(i2c1, I2C_FREQ);
     gpio_set_function(PIN_SDA, GPIO_FUNC_I2C);
     gpio_set_function(PIN_SCL, GPIO_FUNC_I2C);
     gpio_pull_up(PIN_SDA);
     gpio_pull_up(PIN_SCL);
     if (!gpio_get(PIN_SDA) || !gpio_get(PIN_SCL)) {
-        printf("ERROR: I2C bus is not clean\n");
+        info("ERROR: I2C bus is not clean, unplug the controller\n");
```
Comment (przymi) [inline src/bus.c]: ERROR, not INFO

File: src/bus.c [hunk 6]
```diff
@@ -112,22 +113,22 @@ void bus_i2c_io_init_single(uint8_t id) {
     bus_i2c_write(id, I2C_IO_REG_POLARITY+1, 0b11111111);
     bus_i2c_write(id, I2C_IO_REG_PULL,   0b11111111);
     bus_i2c_write(id, I2C_IO_REG_PULL+1, 0b11111111);
-    printf("  IO id=%i ", id);
-    printf("ack=%i ", bus_i2c_acknowledge(id));
-    printf("polarity=%i ", bin(bus_i2c_read_one(id, I2C_IO_REG_POLARITY)));
-    printf("pull=%i\n", bin(bus_i2c_read_one(id, I2C_IO_REG_PULL)));
+    info("  IO id=%i ", id);
+    info("ack=%i ", bus_i2c_acknowledge(id));
+    info("polarity=0x%i ", bin(bus_i2c_read_one(id, I2C_IO_REG_POLARITY)));
```
Comment (przymi) [inline src/bus.c]: So these are now hex?

File: src/config.c [hunk 7]
```diff
@@ -180,43 +192,37 @@ void config_tune(bool direction) {
     config_read(&config);
     int8_t value = direction ? 1 : -1;
     if (config_tune_mode == PROC_TUNE_OS) {
-        config.os_mode = constrain(config.os_mode + value, 0, 2);
-        printf("Tune: OS mode set to preset %i\n", config.os_mode);
```
Comment (przymi) [inline src/config.c:184]: Why no info?

File: src/config.c [hunk 8]
```diff
@@ -180,43 +192,37 @@ void config_tune(bool direction) {
...
-        config_write(&config);
-        profile_pending_reboot = true;
```
Comment (przymi) [inline src/config.c:186]: Why are these no longer set / written?

File: src/config.c [hunk 9]
```diff
@@ -180,43 +192,37 @@ void config_tune(bool direction) {
...
-        hid_allow_communication = false;
+        config_set_protocol(constrain(config.os_mode + value, 0, 2));
     }
-    if (config_tune_mode == PROC_TUNE_SENSITIVITY) {
-        config.sensitivity = constrain(config.sensitivity + value, 0, 2);
-        config_write(&config);
```
Comment (przymi) [inline src/config.c:191]: See above

File: src/config.c [hunk 10]
```diff
@@ -180,43 +192,37 @@ void config_tune(bool direction) {
...
-void config_calibrate_execute(alarm_id_t alarm) {
-    cancel_alarm(alarm);
+void config_factory() {
+    info("NVM: Reset to factory defaults\n");
+    config_write_init();
+    config_print();
+    config_reboot();
+}
+
+void config_calibrate_execute() {
```
Comment (przymi) [inline src/config.c:226]: Where are you taking care of the alarm?

File: src/config.c [hunk 11]
```diff
@@ -226,10 +232,19 @@ void config_calibrate_execute(alarm_id_t alarm) {
 }
 
 void config_calibrate() {
+    logging_set_onloop(false);
+    info("Calibration about to start, leave the controller on a flat surface\n");
     profile_led_lock = true;
     led_shape_all_off();
     led_blink_mask(LED_MASK_LEFT | LED_MASK_RIGHT);
-    add_alarm_in_ms(5000, (alarm_callback_t)config_calibrate_execute, NULL, true);
```
Comment (przymi) [inline src/config.c:232]: Why is this removed?

File: src/config.c [hunk 12]
```diff
@@ -238,24 +253,89 @@ void config_set_pcb_gen(uint8_t gen) {
 
 uint8_t config_get_pcb_gen() {
     if (pcb_gen == 255) {
-        printf("ERROR: PCB gen could not be determined\n");
+        info("ERROR: PCB gen could not be determined\n");
```
Comment (przymi) [inline src/config.c]: ERROR, not INFO

File: src/config.c [hunk 13]
```diff
@@ -238,24 +253,89 @@ void config_set_pcb_gen(uint8_t gen) {
...
         sleep_ms(1000000000);
```
Comment (przymi) [inline src/config.c]: Why is there a sleep here?

File: src/headers/hid.h [hunk 14]
```diff
@@ -198,10 +198,12 @@
...
 #define PROC_CALIBRATE             PROC_INDEX + 20
-#define PROC_BOOTSEL               PROC_INDEX + 21
-#define PROC_THANKS                PROC_INDEX + 22
-#define PROC_MACRO                 PROC_INDEX + 23
-#define PROC_HOME_GAMEPAD          PROC_INDEX + 24
+#define PROC_RESTART               PROC_INDEX + 21
+#define PROC_BOOTSEL               PROC_INDEX + 22
+#define PROC_FACTORY               PROC_INDEX + 23
+#define PROC_THANKS                PROC_INDEX + 24
+#define PROC_MACRO                 PROC_INDEX + 25
```
Comment (przymi) [inline src/headers/hid.h:205]: These will conflict with ADZ and ADZN below

File: src/headers/logging.h [hunk 15]
```diff
@@ -0,0 +1,21 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright (C) 2022, Input Labs Oy.
+
+#pragma once
+
+typedef enum LogLevel_enum {
+    LOG_INFO,
+    LOG_DEBUG,
```
Comment (przymi) [inline src/headers/logging.h:10]: Where are the WARN and ERROR levels?

File: src/headers/tusb_config.h [hunk 16]
```diff
@@ -26,36 +38,36 @@
...
 #define STRING_PRODUCT "Alpakka"
 #define STRING_DEVICE_VERSION "1.0"
-#define STRING_INTERFACE_0 "HID"
-#define STRING_INTERFACE_1 "XINPUT_GENERIC_CONTROLLER"
+#define STRING_HID "HID"
+#define STRING_WEBUSB "WEBUSB"
+#define STRING_XINPUT "XINPUT_GENERIC_CONTROLLER"
 
-#define WCID_VENDOR 0x17
+#define MS_OS_VENDOR 0x17
```
Comment (przymi) [inline src/headers/tusb_config.h:45]: Why this change?

File: src/headers/tusb_config.h [hunk 17]
```diff
@@ -12,12 +12,24 @@
...
 #define CFG_TUSB_MEM_ALIGN __attribute__ ((aligned(4)))
 #define CFG_TUD_ENDPOINT0_SIZE 64
+#define CFG_TUD_VENDOR_RX_BUFSIZE 64
+#define CFG_TUD_VENDOR_TX_BUFSIZE 64
 
 #define CFG_TUD_HID 1
 #define CFG_TUD_CDC 0
 #define CFG_TUD_MSC 0
 #define CFG_TUD_MIDI 0
-#define CFG_TUD_VENDOR 0
```
Comment (przymi) [inline src/headers/tusb_config.h:20]: Why?

File: src/headers/tusb_config.h [hunk 18]
```diff
@@ -67,10 +79,30 @@
...
+        ADDR_WEBUSB_OUT,  /* Address out */\
+        ADDR_WEBUSB_IN,   /* Address in */\
+        64                /* Size */\
+    )
+
 #define DESCRIPTOR_INTERFACE_XINPUT \
     0x09,        /* bLength */\
     0x04,        /* bDescriptorType: interface */\
-    0x01,        /* bInterfaceNumber */\
+    ITF_XINPUT,  /* bInterfaceNumber */\
```
Comment (przymi) [inline src/headers/tusb_config.h:105]: ITF_XINPUT is 2 now

File: src/config.c [hunk 19]
```diff
@@ -238,24 +253,89 @@ void config_set_pcb_gen(uint8_t gen) {
...
+    info("Pico UID: %s\n", pico_id);
+    info("INIT: Config\n");
     config_nvm_t config;
     config_read(&config);
     if (
         config.header != NVM_CONFIG_HEADER ||
         config.config_version != CFG_STRUCT_VERSION
     ) {
-        printf("  config not found or incompatible, writing default instead\n");
+        info("  config not found or incompatible, writing default instead\n");
```
Comment (przymi) [inline src/config.c]: Better ERROR

**Review by przymi — CHANGES_REQUESTED** [review]

File: src/headers/tusb_config.h [hunk 20]
```diff
@@ -85,58 +117,89 @@
...
-    0x20, 0x00,  /* wMaxPacketSize */\
-    0x04         /* bInterval */\
+    0x07,            /* bLength */\
+    0x05,            /* bDescriptorType: endpoint */\
+    ADDR_XINPUT_IN,  /* bEndpointAddress */\
+    0x03,            /* bmAttributes */\
+    0x20, 0x00,      /* wMaxPacketSize */\
+    0x04             /* bInterval */\
 
 #define DESCRIPTOR_ENDPOINT_XINPUT_OUT \
```
Comment (przymi) [inline src/headers/tusb_config.h:127]: Skipping review of rest of this file due to magic

File: src/hid.c [hunk 21]
```diff
@@ -423,11 +426,11 @@ void hid_report() {
         is_tud_ready = true;
         if (!is_tud_ready_logged) {
             is_tud_ready_logged = true;
-            // hid_matrix_reset();
-            printf("USB: tud_ready TRUE\n");
+            info("USB: tud_ready TRUE\n");
```
Comment (przymi) [inline src/hid.c:429]: What is this supposed to do?

File: src/imu.c [hunk 22]
```diff
@@ -179,5 +181,4 @@ void imu_calibrate() {
         offset_accel_1_y,
         offset_accel_1_z
     );
-    printf("Calibration done\n");
```
Comment (przymi) [inline src/imu.c:182]: Why no more notification?

File: src/imu.c [hunk 23]
```diff
@@ -16,7 +16,9 @@
 #include "touch.h"
 #include "hid.h"
 #include "led.h"
+#include "vector.h"
```
Comment (przymi) [inline src/imu.c:19]: Why is this needed now?

File: src/logging.c [hunk 24]
```diff
@@ -0,0 +1,62 @@
...
+#include <stdio.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdarg.h>
+#include "logging.h"
+#include "webusb.h"
+#include "helper.h"
+
+LogLevel logging_level = LOG_INFO;
+bool logging_onloop = false;
```
Comment (przymi) [inline src/logging.c:16]: What is this used for?

File: src/tusb_config.c [hunk 25]
```diff
@@ -55,23 +59,26 @@ uint8_t const *tud_descriptor_device_cb() {
...
 }
 
 uint8_t const *tud_hid_descriptor_report_cb(uint8_t instance) {
-    printf("USB: tud_hid_descriptor_report_cb\n");
+    loguart("USB: tud_hid_descriptor_report_cb\n");
     if (config_get_os_mode() == OS_MODE_GENERIC) return descriptor_report_generic;
     else return descriptor_report_xinput;
 }
 
 const uint16_t *tud_descriptor_string_cb(uint8_t index, uint16_t langid) {
```
Comment (przymi) [inline src/tusb_config.c:78]: What is this supposed to do?

File: src/tusb_config.c [hunk 26]
```diff
@@ -93,21 +100,37 @@ const bool tud_vendor_control_xfer_cb(
...
+    if (
+        request->wIndex == 0x0005 &&
+        request->bRequest == MS_OS_VENDOR &&
+        config_get_os_mode() != OS_MODE_XINPUT_UNIX
+    ) {
+        static uint8_t response[] = {MS_OS_PROPERTIES};
+        return tud_control_xfer(rhport, request, response, sizeof(response));
     }
-    return true;
+    return false;
```
Comment (przymi) [inline src/tusb_config.c:134]: ???

File: src/webusb.c [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+
+char webusb_buffer[WEBUSB_BUFFER_SIZE] = {0,};
+uint16_t webusb_ptr_in = 0;
+uint16_t webusb_ptr_out = 0;
+bool webusb_timedout = false;
+Ctrl_cfg_type webusb_pending_config_give = 0;
+
+Ctrl webusb_ctrl_log() {
+    Ctrl ctrl = {
+        .protocol_version = 1,
```
Comment (przymi) [inline src/webusb.c]: Should be defined in header

File: src/webusb.c [hunk 28]
```diff
@@ -0,0 +1,158 @@
...
+        .device_id = 1,
+        .message_type = LOG
+    };
+    ctrl.len = max(0, min(64-4, webusb_ptr_in - webusb_ptr_out));
```
Comment (przymi) [inline src/webusb.c]: magic!

File: src/webusb.c [hunk 29]
```diff
@@ -0,0 +1,158 @@
...
+    }
+    webusb_ptr_out += ctrl.len;
+    if (webusb_ptr_out >= webusb_ptr_in) {
+        webusb_ptr_in = 0;
+        webusb_ptr_out = 0;
+    }
+    return ctrl;
+}
+
+Ctrl webusb_ctrl_config_give() {
```
Comment (przymi) [inline src/webusb.c]: Should be  notify (or return)

File: src/webusb.c [hunk 30]
```diff
@@ -0,0 +1,158 @@
...
+    Ctrl ctrl = {
+        .protocol_version = 1,
```
Comment (przymi) [inline src/webusb.c]: Define in header

**Review by przymi — CHANGES_REQUESTED** [review]

File: src/bus.c (hunk 6 above)
Comment (marcos-diaz) [reply src/bus.c]: My fault, these are binary so it should be `0b` instead of `0x`.

File: src/config.c (hunk 10 above)
Comment (marcos-diaz) [reply src/config.c:226]: No alarm anymore (because the limitation of not being able to to IRQ inside another IRQ), and because it was not actually needed after all.

File: src/headers/tusb_config.h (hunk 18 above)
Comment (marcos-diaz) [reply src/headers/tusb_config.h:105]: Yes, that's expected since interface #1 is now webusb.

File: src/config.c (hunk 19 above)
Comment (marcos-diaz) [reply src/config.c]: I will make it a warning

File: src/imu.c (hunk 23 above)
Comment (marcos-diaz) [reply src/imu.c:19]: Some functions return `Vector`

File: docs/ctrl_protocol.md (hunk 3 above)
Comment (marcos-diaz) [reply docs/ctrl_protocol.md]: Renamed to `CONFIG_SHARE`

File: docs/ctrl_protocol.md (hunk 4 above)
Comment (marcos-diaz) [reply docs/ctrl_protocol.md]: Added descriptions to all sections

File: docs/ctrl_protocol.md
Comment (marcos-diaz) [reply docs/ctrl_protocol.md:1]: Added sequence diagram

File: src/config.c (hunk 8 above)
Comment (marcos-diaz) [reply src/config.c:186]: The logic was moved as it was to the functions in the bottom of the file

File: src/config.c (hunk 11 above)
Comment (marcos-diaz) [reply src/config.c:232]: It was not really needed, and it was buggy because there cannot be `sleep` calls inside the callback of an IRQ

File: src/config.c (hunk 13 above)
Comment (marcos-diaz) [reply src/config.c]: Added more self-explanatory `exit()` function

File: src/headers/tusb_config.h (hunk 16 above)
Comment (marcos-diaz) [reply src/headers/tusb_config.h:45]: `MS_OS` is a better name.

`WCID` means `Windows Compatible IDs` and it only represent a subset of what is happening there.

File: src/headers/tusb_config.h (hunk 17 above)
Comment (marcos-diaz) [reply src/headers/tusb_config.h:20]: I was not able to make webusb to work without it, which makes sense, in fact I'm surprised xinput was working at all without TinyUSB vendor enabled.

File: src/imu.c (hunk 22 above)
Comment (marcos-diaz) [reply src/imu.c:182]: There is notification at a higher level (the function that wraps both stick and IMU calibrations)
